### PR TITLE
[clang-scan-deps] Fix contention when updating `TrackingStatistic`s in hot code paths in `FileManager`.

### PR DESCRIPTION
**Pull request by aganea**
`FileManager::getDirectoryRef` and `FileManager::getFileRef` are hot code paths in `clang-scan-deps`. In these functions, we update a couple of atomic variables related to printing statistics, which causes contention on high core count machines.

Screenshot 2024-04-10 214123

Screenshot 2024-04-10 214246

Screenshot 2024-04-10 214315

After this PR, we update these variables iff the user wants to print statistics.

On my use case, this saves about 49 sec over 1 min 47 sec of `clang-scan-deps` run time (1 min 47 sec before, 58 sec after). These figures are after applying my suggestion in https://github.com/llvm/llvm-project/pull/88152#issuecomment-2049803229, that is:
```
static bool shouldCacheStatFailures(StringRef Filename) {
  return true;
}
```
Without the above, there's just too much OS noise from the high volume of `status()` calls with regular non-modules C++ code. Tested on Windows with clang-cl.

### CODE REVIEW

**Comment by jansvoboda11** [comment]
I'd like for @vsapsai to chime in.

As an alternative approach: could we turn these into member variables, make them non-atomic and take care to update the stats of the superior `FileManager` whenever an inferior `FileManager` goes out of scope? (e.g. after implicitly building a module)

**Comment by vsapsai** [comment]
The bigger idea is that not enabled stats should be negligibly cheap. As for me, the problem is in that not being true. Let me check how we have expensive stats all the time.

Extra fixes are good but I'm concerned they help only with these specific stats while others might still cause problems in different contexts.

**Comment by aganea** [comment]
I think in the short term @jansvoboda11's suggestion should be good enough.

Bit if we want `Statistics` to be always cheap, we should make them `thread_local` instead, not atomic. `getValue()` could do the "collection" of data over all active, or past threads.  It would also need a mechanism for collecting data when a thread ends through `pthread_key_create/FlsCallback`s.  It would be a bit more involved than what's there currently, but that should fix the issue I'm seeing (and maybe others).

**Comment by aganea** [comment]
> As an alternative approach: could we turn these into member variables, make them non-atomic and take care to update the stats of the superior `FileManager` whenever an inferior `FileManager` goes out of scope? (e.g. after implicitly building a module)

As suggested.

File: clang/lib/Frontend/CompilerInstance.cpp [hunk 1]
```diff
@@ -1293,6 +1293,10 @@ compileModuleImpl(CompilerInstance &ImportingInstance, SourceLocation ImportLoc,
                                             diag::remark_module_build_done)
     << ModuleName;
 
+  // Propagate the statistics to the parent FileManager.
+  if (FrontendOpts.ModulesShareFileManager)
```
Comment (jansvoboda11) [inline clang/lib/Frontend/CompilerInstance.cpp:1297]: Shouldn't this condition be inverted? When `FrontendOpts.ModulesShareFileManager == true`, I'd expect `&ImportingInstance.getFileManager() == &Instance.getFileManager()`. Maybe we can have an assertion in `FileManager::AddStats()` that would catch this.

Comment (aganea) [reply clang/lib/Frontend/CompilerInstance.cpp]: Yes, you're right, fixed!

**Review by jansvoboda11 — APPROVED** [review]
LGTM, thanks!

**Comment by jansvoboda11** [comment]
(Out of interest, what machine are you seeing the contention with?)

**Comment by aganea** [comment]
> LGTM, thanks!

Thanks for the review!

> (Out of interest, what machine are you seeing the contention with?)

It's a ThreadRipper Pro 3975WX 32c/64t running on with Windows 11.

**Comment by MaskRay** [comment]
(BTW: https://discourse.llvm.org/t/hidden-emails-on-github-should-we-do-something-about-it/74223 The recommendation is to avoid the github "hidden email" feature.)

**Comment by aganea** [comment]
Thanks for pointing that out @MaskRay !

**Pull request merged**